homeassistant/components/zwave_js/fan.py

```diff
@@ -206,3 +185,41 @@ def speed_count(self) -> int:
         assert self.speed_configuration
 
         return len(self.speed_configuration)
+
+    def percentage_to_zwave_speed(self, percentage: int) -> int:
+        """Map a percentage to a ZWave speed."""
+        if percentage == 0:
+            return 0
+
+        # Entity should be unavailable if this isn't set
+        assert self.speed_configuration
+
+        # Since the percentage steps are computed with rounding, we have to
+        # search to find the appropriate speed.
+        for speed_limit in self.speed_configuration:
+            step_percentage = self.zwave_speed_to_percentage(speed_limit)
+            if percentage <= step_percentage:
+                return speed_limit
+
+        # This shouldn't actually happen; the last entry in
+        # `self.speed_configuration` should map to 100%.
+        return 100
+
+    def zwave_speed_to_percentage(self, zwave_speed: int) -> int:
+        """Convert a Zwave speed to a percentage."""
+        if zwave_speed == 0:
+            return 0
+
+        # Entity should be unavailable if this isn't set
+        assert self.speed_configuration
+
+        percentage = 0.0
+        for speed_limit in self.speed_configuration:
+            percentage += self.percentage_step
+            if zwave_speed <= speed_limit:
+                break
+
+        # This choice of rounding function is to provide consistency with how
```

Comment:
@MartinHjelmare Can you give me a sanity-check on whether this is the right way to go here?

As the comment notes, for 3-speed fans, the slider in the UI gets steps at 33, 67, and 100.  So, assuming that this is all well and good, I made this conversion line up with that so that 33=low, 67=medium, and 100=high.

But the tricky thing is that HA has traditionally used 33, *66*, and 100 as the steps for low, medium, and high.  And there's a bit of conversion logic in the base fan component that assumes as much:

https://github.com/home-assistant/core/blob/6a93f5b7ad1fc751be8ac3a57aea10f9858a7b6e/homeassistant/components/fan/__init__.py#L581-L588

But my understanding is that those speed constants are deprecated, so maybe it's not worth changing for them?

Reply:
Those old speed settings are indeed deprecated and not relevant for us.

Reply:
Thanks for confirming!